state/storage.go

```diff
@@ -182,13 +182,22 @@ func (s *Storage) MetadataForAllRooms(txn *sqlx.Tx, tempTableName string, result
 	}
 
 	// work out latest timestamps
-	events, err := s.accumulator.eventsTable.selectLatestEventInAllRooms(txn)
+	events, err := s.accumulator.eventsTable.selectLatestEventByTypeInAllRooms(txn)
 	if err != nil {
 		return err
 	}
 	for _, ev := range events {
-		metadata := result[ev.RoomID]
+		metadata, ok := result[ev.RoomID]
 		metadata.LastMessageTimestamp = gjson.ParseBytes(ev.JSON).Get("origin_server_ts").Uint()
+		if !ok {
+			metadata = *internal.NewRoomMetadata(ev.RoomID)
+		}
```

Comment:
I'm not sure if this is needed. The comment below makes it sound like the metadata may not be present in the map.  If so, I don't want to write to a `LatestEventsByType` which is a nil map, because AFAIK that panics.

If that reasoning is sound, why did we get away with this before? internal.RoomMetadata has a ChildSpaceRooms map which is nil unless you use the NewRoomMetadata helper?